Release/step1

yanghun0070 opened the pull request:
요구사항 추가 및 도메인 객체 생성

yangjeehan committed:
feat($location) : add requirement

Added project requirement

It is initial requirement.

Commit:
fix(etc) : add glossary

add glossary shopping-gogo

Commit:
create(domain)

create shopping domain

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/cart/domain/Cart.java:
```diff
@@ -0,0 +1,19 @@
+    public Cart(String id, User user, List<Product> productList) {
+        this.id = id;
+        this.user = user;
+        this.productList = productList;
```
서로 다른 AGGREGATE 간에 직접 참조할 때 발생할 수 있는 가장 큰 문제는 편리함을 오용할 수 있습니다. ID 참조를 사용하는 것은 어떨까요?

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/product/domain/Product.java:
```diff
@@ -0,0 +1,19 @@
+    private String id;
+    private String name;
+    private int lowestPrice;
+    private int highestPrice;
```
`lowestPrice`와 `highestPrice`는 `Price`라는 VALUE OBJECT로 분리해보는 것은 어떨까요?

yanghun0070 replied:
넵, 강사님 말씀해주신데로 변경하였습니다 ^^;;
지금 저희가 오해를 하여, 지금 커밋 리뷰 댓글 수정중에 있네여 ㅠㅠ

wotjd243 reviewed (changes requested):
추가적인 설계와 구현이 필요해 보입니다.
피드백 남겼으니 확인하고 개선할 부분 있으면 반영해 보세요.
코드 개선 시, 모델링도 함께 업데이트되어야 하는 점 잊지 마세요.

Commit:
create(user domain)

유저 도메인 및 주문 도메인 생성

Commit:
feature(userService)

회원인지 아닌지 판단하는 기능 추가 , 코드 리뷰( password 삭제)

yangjeehan committed:
add payment and shipping

Commit:
장바구니, 상품 도메인 설계

Commit:
Merge commit '94b578e5ffe07d9fea6bacee9bceeef6b8d25447' into shopping-go-go

Commit:
Merge commit 'a11cbf2e597a4be7d25e736c12e7f8a1cf1bb859' into shopping-go-go

# Conflicts:
#	src/main/java/io/github/wotjd243/shoppinggogo/cart/domain/Cart.java
#	src/main/java/io/github/wotjd243/shoppinggogo/product/domain/Product.java

Commit:
수정사항 커밋

Commit:
로그인 함수 수정
Dummy 데이터 수정

Commit:
에러 사항 수정

yangjeehan committed:
feat() : changes

Added Payment , Shipping and Seller Aggregation
And change UserId Type to String

yanghun0070 committed:
Cart, Product Domain 설계 및 테스트 코드 작성

yanghun0070 committed:
Merge commit '3d694c0bc4bc28c82e5e93b1cd4bee97dc9cd6ec' into shopping-go-go

# Conflicts:
#	build.gradle
#	src/main/java/io/github/wotjd243/shoppinggogo/user/domain/UserRepository.java

Commit:
feature( order Domain)

Commit:
feature(ordeerDomain)

Commit:
Merge branch 'shopping-go-go' of https://github.com/yanghun0070/ddd-project into shopping-go-go

yangjeehan committed:
fix(CashPayment.java) : change

fix CashPayment

yanghun0070 committed:
Lombok 커밋

yanghun0070 committed:
Merge commit 'd248cdca1a4c60fc6e68b08fe08a21e684e92b8d' into shopping-go-go

# Conflicts:
#	src/main/java/io/github/wotjd243/shoppinggogo/cart/domain/Cart.java

yanghun0070 committed:
패키지 위치가 달라서 에러나는 거 수정

Commit:
Merge branch 'shopping-go-go' of https://github.com/yanghun0070/ddd-project into shopping-go-go

yangjeehan committed:
fix(Shipping) :

fix Shipping : delete mall, destination column

yangjeehan committed:
Merge branch 'shopping-go-go' of https://github.com/yanghun0070/ddd-project into shopping-go-go

Commit:
주문 도메인 추가

yanghun0070 committed:
Cart checked 로직 수정

yanghun0070 committed:
Merge branch 'shopping-go-go' of https://github.com/yanghun0070/ddd-project into shopping-go-go

Commit:
Merge branch 'shopping-go-go' of https://github.com/yanghun0070/ddd-project into shopping-go-go

Commit:
feature(createDomain)

Commit:
feature(change id type  string to long)

yanghun0070 committed:
Merge branch 'shopping-go-go' into release/step1

yanghun0070 committed:
Cart 에 담긴 제품 ID 목록을 찾는 함수 구현

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/cart/domain/Cart.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Data
```
Lombok의 `@Data`는 `@ToString`, `@EqualsAndHashCode`, `@Getter`가 모든 멤버 변수에, `@Setter`는 모든 non-final 멤버 변수에 사용되며 `@RequiredArgsConstructor`가 사용되는 매우 강력한 어노테이션입니다. 아래의 링크를 참고하여 가급적 ENTITY에는 사용하지 않는 것을 추천합니다.
https://www.popit.kr/%EC%8B%A4%EB%AC%B4%EC%97%90%EC%84%9C-lombok-%EC%82%AC%EC%9A%A9%EB%B2%95

yanghun0070 replied:
넵, 수정하겠습니다. ^^~!

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/cart/domain/Cart.java:
```diff
@@ -0,0 +1,61 @@
+    private Long id; //카트 키
+    private Long userId; //유저 키
+    private List<Product> products; //제품 목록
+    private List<Long> checkedProductIds; //체크되어진 제품 ID 목록
```
"체크된 제품 ID 목록"이라는 것은 요구 사항과 용어 사전, 모델링에서 알 수가 없네요. `Cart` 입장에서 "체크된 제품 ID 목록"이라는 것은 데이터베이스에 저장하고 관리해야 할 상태 값인가요? `matchByCheckedProductIds()`를 동작시키기 위한 용도라면 상태 값으로 안 가져도 될 것 같아요.

yanghun0070 replied:
이부분은 잘못 생각해서 생긴 부분입니다~! 
지적 감사합니다 수정하였습니다.

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/cart/domain/Cart.java:
```diff
@@ -0,0 +1,61 @@
+    @Getter
+    private Long id; //카트 키
+    private Long userId; //유저 키
+    private List<Product> products; //제품 목록
```
이전 리뷰에서 코멘트를 남겼었습니다. 서로 다른 AGGREGATE 간에 직접 참조할 때 발생할 수 있는 가장 큰 문제는 편리함을 오용할 수 있습니다. ID 참조를 사용하는 것은 어떨까요?

yanghun0070 replied:
넵, 저건 말씀하신데로 Id로 변경하였습니다. ~
다음 step2 에서 확인하실 수 있을꺼예요~!

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/cart/infra/CartRepository.java:
```diff
@@ -0,0 +1,18 @@
+     * @param userId 유저 ID
+     * @return Cart 에 담긴 제품 목록
+     */
+    public List<Product> selectProductsToCart(Long userId);
```
REPOSITORY가 AGGREGATE인 `Cart`를 조회해온 다음 `getProducts()`를 이용하는 것은 어떨까요? 도메인 주도 설계를 잠시 잊고 '우아한 코드'에 대해서 생각해보면 좋을 것 같아요.
https://youtu.be/MDZ8wUi-jzU

yanghun0070 replied:
좋은 자료네요 감사합니다. :) 👍 👍 👍

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/payment/domain/Amount.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Amount {
+
+    public static final int MINIMUM_NUMBER = 1;
```
매직넘버를 상수로 관리 👍

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/payment/domain/Amount.java:
```diff
@@ -0,0 +1,42 @@
+
+    public static final int MINIMUM_NUMBER = 1;
+    // 1억이상 결제 x
+    public static final int MAXIMUM_NUMBER = 100000000;
```
Java7부터는 `100_000_000`처럼 숫자에 underscore("_")를 사용하여 가독성을 높일 수 있습니다.

yangjeehan replied:
말씀하신대로 수정했습니다

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/payment/domain/Payment.java:
```diff
@@ -0,0 +1,24 @@
+    private String userId; //유저
+
+    public Payment(Date paymentDate, Long id, String orderId, Amount amountForPayment, String userId) {
+        this.paymentDate = paymentDate;
```
혹시 Date 클래스의 위험성에 대해서 알고 계시나요? 아래의 글을 참고하여 외부에 의한 변조가 일어날 수 없도록 변경해보면 좋을 것 같아요.
https://d2.naver.com/helloworld/645609

yangjeehan replied:
Date 클래스에서 setMethod을 가지고 있어 문제가 발생할 수 있다는 점을 리마인드 했습니다 감사합니다

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/shipping/domain/GeneralBusinessNumber.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    private void validate(final String businessNum) {
+        String ePattern = "^\\d{3}-\\d{2}-\\d{4}$";
```
매직넘버는 상수로 관리하는 것이 그 의미를 이해하는 데 도움이 될 것 같아요. 그리고 `String`을 패턴 매칭을 위해 `Pattern`으로 바꾸는 것은 꽤나 비싼 행위이므로 미리 `Pattern`으로 만들어 사용하는 것은 어떨까요?

yangjeehan replied:
말씀하신바가  아래와 같이 구현 하는게 맞나요 ? 
    private final Pattern BUSINESS_NUMBER_PATTERN = Pattern.compile("^\\d{3}-\\d{2}-\\d{4}$");

    public String businessNum;

    private GeneralBusinessNumber(final String businessNum) {
        validate(businessNum);
        this.businessNum = businessNum;
    }

    private void validate(final String businessNum) {
       if( !BUSINESS_NUMBER_PATTERN.matcher(businessNum).matches() ) {
            throw new IllegalArgumentException();
        }
    }

wotjd243 replied:
네, 맞습니다. 👍

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/shipping/domain/GeneralEmail.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    private void validate(final String email) {
+        String ePattern = "^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$";
```
마찬가지로 매직넘버는 상수로 관리하는 것이 그 의미를 이해하는 데 도움이 될 것 같아요.

wotjd243 commented on README.md:
```diff
@@ -1 +1,64 @@
+| 배송비 | shipping fee |장바구니 목록에서, 같은 쇼핑몰인지 아닌지에 따라서,배송비가 무료 또는 2500원이 붙는다.
+
+ ###모델링
+ 유저 엔
```
현재까지 구현된 것을 모델링으로 표현해보는 것은 어떨까요?

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/user/ui/UserController.java:
```diff
@@ -0,0 +1,26 @@
+public class UserController {
+
+    @Autowired
+    UserService userService;
```
Spring에서 필드 주입을 권장하지 않는 이유가 무엇일까요?

1. 단일 책임의 원칙(Single Responsibility Principle)을 위반할 수 있다.
2. 의존성이 클래스 내부로 숨는다.
3. 인스턴스화가 어려워 테스트하기 어렵다.
4. final을 쓸 수 없어 불변성이 보장되지 않는다.

위의 이유로 스프링 공식 문서에서도 생성자 주입을 권장하고 있습니다.

yanghun0070 replied:
이부분에 대해서는 잘 몰랐던 부분이네요.. ^^; 👍

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/cart/domain/Cart.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.stream.Collectors;
+
+@Data
+public class Cart {
```
"장바구니 목록에서 두 개 이상의 상품의 쇼핑몰이 같다면, 배송비는 무료로 배송된다.", "장바구니의 상품들의 쇼핑몰 이름 중복이 없다면, 상품 개별 2500원이 추가 결제된다."와 같은 요구 사항은 구현이 필요해 보입니다.

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/order/domain/Order.java:
```diff
@@ -0,0 +1,30 @@
+    public Order(Long userId, Address address, Phone phone, List<Long> products) {
+        this.userId = userId;
+        this.destination = address;
+        this.orderProducts = products;
```
새로운 리스트 컬렉션에 담아 외부에 의한 변조를 막는 것은 어떨까요?

yanghun0070 replied:
넵, 다음 step2 에 변경될 거예요~

wotjd243 commented on src/main/java/io/github/wotjd243/shoppinggogo/product/domain/Product.java:
```diff
@@ -0,0 +1,45 @@
+    private String name; //상품 명
+    private Category category; //상품 카테고리
+    private ProductImage productImage; //제품 이미지
+    private List<PriceRecord> prices;
```
`PriceRecord`가 컬렉션으로 관리되는 이유가 무엇일까요? 쇼핑몰마다 각각의 최저가와 최고가가 있다는 뜻인가요? 요구 사항과 용어 사전, 모델링을 보아도 알 수 없어 궁금하네요.

yanghun0070 replied:
넵,  용어사전, 모델링이 업데이트가 되지 않았네요.
초반에 목적은 `PriceRecord` **업체별 가격 기록**들을 목록을 API에서 확인할 수 있을 거라고 생각해서 구현했습니다.
네이버API 에서 제공해주지 않아서 다시 변경사항을 진행할 예정입니다~

wotjd243 commented on src/test/java/io/github/wotjd243/shoppinggogo/cart/domain/CartTest.java:
```diff
@@ -0,0 +1,9 @@
+
+import org.junit.Test;
+
+public class CartTest {
```
중요한 AGGREGATE 중 하나인 `Cart`에 대한 테스트 코드가 빠져있네요. 테스트 코드를 작성해보면 어떨까요?

yanghun0070 replied:
넵, 이부분 관련해서 다음 step2 에 진행 하겠습니다. ^^~!

wotjd243 reviewed (approved):
지난 리뷰 때보다 많은 기능이 구현되었네요. 👍 
하지만 아직 사용자가 장바구니에 담은 상품이 어느 쇼핑몰의 상품인지 구분 등의
구현되지 않은 요구 사항이 눈에 보이며, 도메인 계층에 대한 테스트 코드가 많이 빠져있습니다.
요구 사항을 테스트 코드로 작성해보면 혹시 빠진 구현이 없는지 확인할 수 있으며,
다른 개발자에게는 구현 코드에 대한 도메인 지식도 함께 제공해 줄 수 있습니다.
step2에 해당하는 내용은 따로 코멘트하지 않았으며, 몇 가지 생각거리는 따로 남겨두었으니 의논해보세요.
빠른 진행을 위해 Merge 하겠습니다!
